In [1]:
import vaex

In [2]:
import pandas as pd
import numpy as np

In [3]:
flight = pd.read_parquet("C:/Users/HP/Desktop/Programming/Projects/Project 4- Flight/Dataset/Flight_Delay.parquet")

In [4]:
flight.head()

Year  Month  DayofMonth  FlightDate Marketing_Airline_Network  \
1  2018      1          15  2018-01-15                        UA   
2  2018      1          16  2018-01-16                        UA   
3  2018      1          17  2018-01-17                        UA   
4  2018      1          18  2018-01-18                        UA   
6  2018      1          20  2018-01-20                        UA   

  OriginCityName    DestCityName  CRSDepTime  DepTime  DepDelay  ...  \
1     Newark, NJ  Charleston, SC        1845   1928.0      43.0  ...   
2     Newark, NJ  Charleston, SC        1835   1956.0      81.0  ...   
3     Newark, NJ  Charleston, SC        1835   1836.0       1.0  ...   
4     Newark, NJ  Charleston, SC        1845   1844.0      -1.0  ...   
6     Newark, NJ  Charleston, SC        1835   1829.0      -6.0  ...   

   CRSElapsedTime  ActualElapsedTime  AirTime  Distance  DistanceGroup  \
1           143.0              141.0    102.0     628.0              3   
2           143.0              131.0    108.0     628.0              3   
3           143.0              138.0    113.0     628.0              3   
4           143.0              134.0     92.0     628.0              3   
6           143.0              117.0     89.0     628.0              3   

   CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
1          41.0           0.0       0.0            0.0                0.0  
2          67.0           0.0       0.0            0.0                2.0  
3           0.0           0.0       0.0            0.0                0.0  
4           0.0           0.0       0.0            0.0                0.0  
6           0.0           0.0       0.0            0.0                0.0  

[5 rows x 29 columns]

In [5]:
flight.shape

(30132672, 29)

In [6]:
flight_vaex = vaex.from_pandas(flight)

In [7]:
type(flight_vaex)

vaex.dataframe.DataFrameLocal

In [8]:
flight_vaex.head()

#    Year    Month    DayofMonth  FlightDate    Marketing_Airline_Network    OriginCityName    DestCityName      CRSDepTime    DepTime    DepDelay    DepDelayMinutes    TaxiOut    WheelsOff    WheelsOn    TaxiIn    CRSArrTime    ArrTime    ArrDelay    ArrDelayMinutes    CRSElapsedTime    ActualElapsedTime    AirTime    Distance    DistanceGroup    CarrierDelay    WeatherDelay    NASDelay    SecurityDelay    LateAircraftDelay
  0    2018        1            15  2018-01-15    UA                           Newark, NJ        Charleston, SC          1845       1928          43                 43         35         2003        2145         4          2108       2149          41                 41               143                  141        102         628                3              41               0           0                0                    0
  1    2018        1            16  2018-01-16    UA                           Newark, NJ        Charleston, SC          1835       1956          81                 81         18         2014        2202         5          2058       2207          69                 69               143                  131        108         628                3              67               0           0                0                    2
  2    2018        1            17  2018-01-17    UA                           Newark, NJ        Charleston, SC          1835       1836           1                  1         19         1855        2048         6          2058       2054          -4                  0               143                  138        113         628                3               0               0           0                0                    0
  3    2018        1            18  2018-01-18    UA                           Newark, NJ        Charleston, SC          1845       1844          -1                  0         36         1920        2052         6          2108       2058         -10                  0               143                  134         92         628                3               0               0           0                0                    0
  4    2018        1            20  2018-01-20    UA                           Newark, NJ        Charleston, SC          1835       1829          -6                  0         22         1851        2020         6          2058       2026         -32                  0               143                  117         89         628                3               0               0           0                0                    0
  5    2018        1            21  2018-01-21    UA                           Newark, NJ        Charleston, SC          1835       1825         -10                  0         44         1909        2039         5          2058       2044         -14                  0               143                  139         90         628                3               0               0           0                0                    0
  6    2018        1            22  2018-01-22    UA                           Newark, NJ        Charleston, SC          1845       1847           2                  2         41         1928        2118        10          2108       2128          20                 20               143                  161        110         628                3               2               0          18                0                    0
  7    2018        1            23  2018-01-23    UA                           Newark, NJ        Charleston, SC          1835       1823         -12                  0         34         1857        2103        11          2058       2114          16                 16               143                  171        126         628                3               0               0          16                0                    0
  8    2018        1            25  2018-01-25    UA                           Newark, NJ        Charleston, SC          18

In [9]:
flight_vaex.info()

In [10]:
flight_vaex.describe()

Year              Month          DayofMonth  \
data_type               int64              int64               int64   
count                30132672           30132672            30132672   
NA                          0                  0                   0   
mean       2020.0914400156746  6.297564417785452  15.722973754202748   
std                  1.602171           3.487835            8.770224   
min                      2018                  1                   1   
max                      2023                 12                  31   

          FlightDate Marketing_Airline_Network OriginCityName DestCityName  \
data_type     string                    string         string       string   
count       30132672                  30132672       30132672     30132672   
NA                 0                         0              0            0   
mean              --                        --             --           --   
std               --                        --             --           --   
min               --                        --             --           --   
max               --                        --             --           --   

                   CRSDepTime             DepTime            DepDelay  ...  \
data_type               int64             float64             float64  ...   
count                30132672            30132672            30132672  ...   
NA                          0                   0                   0  ...   
mean       1322.7355737652472  1327.2451506789707  10.611442025453302  ...   
std                484.012482          498.389035           51.982926  ...   
min                         1                 1.0              -342.0  ...   
max                      2359              2400.0              7223.0  ...   

               CRSElapsedTime  ActualElapsedTime             AirTime  \
data_type             float64            float64             float64   
count                30132672           30132672            30132672   
NA                          0                  0                   0   
mean       138.86781507461401  132.0091763186484  108.05376035022715   
std                 70.961497           70.61227           68.726209   
min                    -292.0                4.0             -1412.0   
max                    1645.0             1604.0              1557.0   

                    Distance      DistanceGroup       CarrierDelay  \
data_type            float64              int64            float64   
count               30132672           30132672           30132672   
NA                         0                  0                  0   
mean       776.4743040378231  3.580302735847654  5.050472623204474   
std               582.331462           2.291205          33.700239   
min                     16.0                  1                0.0   
max                   5812.0                 11             7232.0   

                 WeatherDelay            NASDelay         SecurityDelay  \
data_type             float64             float64               float64   
count                30132672            30132672              30132672   
NA                          0                   0                     0   
mean       0.8417200107577583  2.9941116738668248  0.027410878132546626   
std                  15.09353           16.642516              1.650743   
min                       0.0                 0.0                   0.0   
max                    2900.0              1848.0                1460.0   

           LateAircraftDelay  
data_type            float64  
count               30132672  
NA                         0  
mean       5.522570052864877  
std                27.517373  
min                      0.0  
max                   2962.0  

[7 rows x 29 columns]

In [11]:
# no null values

In [12]:
count_na = []
for col in flight_vaex.column_names:
    count_na.append(flight_vaex[col].isna().sum().item())
na = pd.Series(data=count_na, index=flight_vaex.column_names)

In [13]:
na

Year                         0
Month                        0
DayofMonth                   0
FlightDate                   0
Marketing_Airline_Network    0
OriginCityName               0
DestCityName                 0
CRSDepTime                   0
DepTime                      0
DepDelay                     0
DepDelayMinutes              0
TaxiOut                      0
WheelsOff                    0
WheelsOn                     0
TaxiIn                       0
CRSArrTime                   0
ArrTime                      0
ArrDelay                     0
ArrDelayMinutes              0
CRSElapsedTime               0
ActualElapsedTime            0
AirTime                      0
Distance                     0
DistanceGroup                0
CarrierDelay                 0
WeatherDelay                 0
NASDelay                     0
SecurityDelay                0
LateAircraftDelay            0
dtype: int64

In [14]:
# Year Column

In [15]:
flight_vaex.Year.unique()

[2021, 2022, 2018, 2020, 2023, 2019]

In [16]:
# Marketing Airline network

In [17]:
flight_vaex.Marketing_Airline_Network.unique()

['F9', 'DL', 'AA', 'HA', 'WN', 'B6', 'VX', 'AS', 'G4', 'NK', 'UA']

In [18]:
# Origin City Name

In [19]:
flight_vaex.OriginCityName.unique()

['Tulsa, OK',
 'St. Louis, MO',
 'Burbank, CA',
 'St. George, UT',
 'Traverse City, MI',
 'Valparaiso, FL',
 'Las Vegas, NV',
 'Kalispell, MT',
 'Hilton Head, SC',
 'Gulfport/Biloxi, MS',
 'Panama City, FL',
 'Raleigh/Durham, NC',
 'Rock Springs, WY',
 'Champaign/Urbana, IL',
 'Staunton, VA',
 'Norfolk, VA',
 'Presque Isle/Houlton, ME',
 'Orlando, FL',
 'Fayetteville, AR',
 'Oklahoma City, OK',
 'Bemidji, MN',
 'Laramie, WY',
 'Portsmouth, NH',
 'Nantucket, MA',
 'Kansas City, MO',
 'Albany, NY',
 'Lewiston, ID',
 'Quincy, IL',
 'Kalamazoo, MI',
 'Gunnison, CO',
 'Flint, MI',
 'Dubuque, IA',
 'Sanford, FL',
 'Myrtle Beach, SC',
 'Anchorage, AK',
 'Sitka, AK',
 'Hyannis, MA',
 'Madison, WI',
 'Boise, ID',
 'Amarillo, TX',
 'Yakutat, AK',
 'Nome, AK',
 'CONCORD, NC',
 'Asheville, NC',
 'Eagle, CO',
 'El Paso, TX',
 'Pocatello, ID',
 'Iron Mountain/Kingsfd, MI',
 'Alexandria, LA',
 'Charleston, SC',
 'Omaha, NE',
 'South Bend, IN',
 'Grand Forks, ND',
 'Milwaukee, WI',
 'Bismarck/Mandan, 

In [20]:
# Destination City Name

In [21]:
flight_vaex.DestCityName.unique()

['Tulsa, OK',
 'St. Louis, MO',
 'Burbank, CA',
 'St. George, UT',
 'Traverse City, MI',
 'Valparaiso, FL',
 'Las Vegas, NV',
 'Kalispell, MT',
 'Hilton Head, SC',
 'Panama City, FL',
 'Gulfport/Biloxi, MS',
 'Raleigh/Durham, NC',
 'Rock Springs, WY',
 'Champaign/Urbana, IL',
 'Staunton, VA',
 'Norfolk, VA',
 'Presque Isle/Houlton, ME',
 'Fayetteville, AR',
 'Oklahoma City, OK',
 'Orlando, FL',
 'Bemidji, MN',
 'Laramie, WY',
 'Portsmouth, NH',
 'Nantucket, MA',
 'Kansas City, MO',
 'Albany, NY',
 'Lewiston, ID',
 'Quincy, IL',
 'Kalamazoo, MI',
 'Gunnison, CO',
 'Flint, MI',
 'Dubuque, IA',
 'Sanford, FL',
 'Myrtle Beach, SC',
 'Anchorage, AK',
 'Sitka, AK',
 'Hyannis, MA',
 'Amarillo, TX',
 'Madison, WI',
 'Boise, ID',
 'Yakutat, AK',
 'Nome, AK',
 'CONCORD, NC',
 'Asheville, NC',
 'Eagle, CO',
 'El Paso, TX',
 'Pocatello, ID',
 'Iron Mountain/Kingsfd, MI',
 'Alexandria, LA',
 'Charleston, SC',
 'Omaha, NE',
 'South Bend, IN',
 'Grand Forks, ND',
 'Milwaukee, WI',
 'Bismarck/Mandan, 

In [22]:
# CRSDep Time, CRS(scheduled) Departure Time (local time: hhmm)

In [23]:
flight_vaex.CRSDepTime

Expression = CRSDepTime
Length: 30,132,672 dtype: int64 (column)
----------------------------------------
       0  1845
       1  1835
       2  1835
       3  1845
       4  1835
     ...      
30132667  1234
30132668  1104
30132669  1249
30132670   751
30132671   635

In [24]:
#  Dep  Time, Actual Departure Time (local time: hhmm)

In [25]:
flight_vaex.DepTime

Expression = DepTime
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  1928
       1  1956
       2  1836
       3  1844
       4  1829
     ...      
30132667  1230
30132668  1100
30132669  1247
30132670   739
30132671   624

In [26]:
# DepDelay (Difference in minutes between scheduled and actual departure time. Early departures show negative numbers.)

In [27]:
flight_vaex.DepDelay

Expression = DepDelay
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0   43
       1   81
       2    1
       3   -1
       4   -6
     ...     
30132667   -4
30132668   -4
30132669   -2
30132670  -12
30132671  -11

In [28]:
# DepDelayMinutes (Difference in minutes between scheduled and actual departure time. Early departures set to 0.)

In [29]:
flight_vaex.DepDelayMinutes

Expression = DepDelayMinutes
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  43
       1  81
       2   1
       3   0
       4   0
    ...     
30132667   0
30132668   0
30132669   0
30132670   0
30132671   0

In [30]:
# Taxi out (Taxi Out Time refers to the duration an aircraft spends taxiing from the gate or terminal to the runway before taking off, in Minutes)

In [31]:
flight_vaex.TaxiOut

Expression = TaxiOut
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  35
       1  18
       2  19
       3  36
       4  22
    ...     
30132667  19
30132668  18
30132669  17
30132670  14
30132671  11

In [32]:
# WheelsOff ("Wheels-Off Time" refers to the exact moment when an aircraft's wheels leave the ground during takeoff. (local time: hhmm))

In [33]:
flight_vaex.WheelsOff

Expression = WheelsOff
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  2003
       1  2014
       2  1855
       3  1920
       4  1851
     ...      
30132667  1249
30132668  1118
30132669  1304
30132670   753
30132671   635

In [34]:
# WheelsOn ("Wheels-On Time" refers to the exact moment when an aircraft's wheels make contact with the runway during landing. (local time: hhmm))

In [35]:
flight_vaex.WheelsOn

Expression = WheelsOn
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  2145
       1  2202
       2  2048
       3  2052
       4  2020
     ...      
30132667  1426
30132668  1340
30132669  1808
30132670  1017
30132671   836

In [36]:
# Taxi in (Taxi In Time is the duration an aircraft spends taxiing from the runway to the gate or terminal after landing, in Minutes)

In [37]:
flight_vaex.TaxiIn

Expression = TaxiIn
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0   4
       1   5
       2   6
       3   6
       4   6
    ...     
30132667   5
30132668   8
30132669  16
30132670  20
30132671   8

In [38]:
# CRSARRtime (CRS(scheduled) Arrival Time (local time: hhmm))

In [39]:
flight_vaex.CRSArrTime

Expression = CRSArrTime
Length: 30,132,672 dtype: int64 (column)
----------------------------------------
       0  2108
       1  2058
       2  2058
       3  2108
       4  2058
     ...      
30132667  1452
30132668  1412
30132669  1829
30132670  1100
30132671   855

In [40]:
# ArrTime (Actual Arrival Time (local time: hhmm))

In [41]:
flight_vaex.ArrTime

Expression = ArrTime
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  2149
       1  2207
       2  2054
       3  2058
       4  2026
     ...      
30132667  1431
30132668  1348
30132669  1824
30132670  1037
30132671   844

In [42]:
# ArrDelay (Difference in minutes between scheduled and actual arrival time. Early arrivals show negative numbers.)

In [43]:
flight_vaex.ArrDelay

Expression = ArrDelay
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0   41
       1   69
       2   -4
       3  -10
       4  -32
     ...     
30132667  -21
30132668  -24
30132669   -5
30132670  -23
30132671  -11

In [44]:
# ArrDelayMinutes (Difference in minutes between scheduled and actual arrival time. Early arrivals set to 0.)

In [45]:
flight_vaex.ArrDelayMinutes

Expression = ArrDelayMinutes
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  41
       1  69
       2   0
       3   0
       4   0
    ...     
30132667   0
30132668   0
30132669   0
30132670   0
30132671   0

In [46]:
# CRSElapsedTime (CRS(scheduled) Elapsed Time of Flight, in Minutes)

In [47]:
flight_vaex.CRSElapsedTime

Expression = CRSElapsedTime
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  143
       1  143
       2  143
       3  143
       4  143
     ...     
30132667  258
30132668  128
30132669  220
30132670  189
30132671  320

In [48]:
# Actual Elapsed time (Elapsed Time of Flight, in Minutes)

In [49]:
flight_vaex.ActualElapsedTime

Expression = ActualElapsedTime
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  141
       1  131
       2  138
       3  134
       4  117
     ...     
30132667  241
30132668  108
30132669  217
30132670  178
30132671  320

In [50]:
# AirTime (Flight Time, in Minutes)

In [51]:
flight_vaex.AirTime

Expression = AirTime
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  102
       1  108
       2  113
       3   92
       4   89
     ...     
30132667  217
30132668   82
30132669  184
30132670  144
30132671  301

In [52]:
# Distance (Distance between airports (miles))

In [53]:
flight_vaex.Distance

Expression = Distance
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0   628
       1   628
       2   628
       3   628
       4   628
     ...      
30132667  1635
30132668   678
30132669  1514
30132670  1085
30132671  2158

In [54]:
# DistanceGroup (Distance Intervals, every 250 Miles, for Flight Segment)

In [55]:
flight_vaex.DistanceGroup

Expression = DistanceGroup
Length: 30,132,672 dtype: int64 (column)
----------------------------------------
       0  3
       1  3
       2  3
       3  3
       4  3
    ...    
30132667  7
30132668  3
30132669  7
30132670  5
30132671  9

In [56]:
# CarrierDelay (Carrier Delay, in Minutes)

In [57]:
flight_vaex.CarrierDelay

Expression = CarrierDelay
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  41
       1  67
       2   0
       3   0
       4   0
    ...     
30132667   0
30132668   0
30132669   0
30132670   0
30132671   0

In [58]:
flight_vaex.CarrierDelay.unique()

[0.0,
 210.0,
 217.0,
 313.0,
 191.0,
 222.0,
 433.0,
 711.0,
 542.0,
 926.0,
 255.0,
 1100.0,
 703.0,
 565.0,
 877.0,
 536.0,
 766.0,
 1105.0,
 1202.0,
 1050.0,
 1199.0,
 1704.0,
 1494.0,
 1341.0,
 1856.0,
 2294.0,
 30.0,
 503.0,
 157.0,
 62.0,
 145.0,
 43.0,
 50.0,
 68.0,
 806.0,
 1179.0,
 1454.0,
 1009.0,
 1168.0,
 432.0,
 1221.0,
 678.0,
 1026.0,
 1204.0,
 2522.0,
 1468.0,
 2365.0,
 2059.0,
 6.0,
 914.0,
 461.0,
 473.0,
 343.0,
 245.0,
 566.0,
 1223.0,
 527.0,
 1205.0,
 755.0,
 1402.0,
 1230.0,
 1317.0,
 1383.0,
 1614.0,
 1701.0,
 2090.0,
 45.0,
 66.0,
 141.0,
 356.0,
 173.0,
 582.0,
 435.0,
 870.0,
 949.0,
 1353.0,
 1363.0,
 1521.0,
 1346.0,
 1792.0,
 1716.0,
 1527.0,
 2240.0,
 166.0,
 349.0,
 69.0,
 815.0,
 532.0,
 545.0,
 1022.0,
 890.0,
 1276.0,
 1403.0,
 1251.0,
 1233.0,
 1249.0,
 1702.0,
 1278.0,
 5324.0,
 2018.0,
 1142.0,
 75.0,
 358.0,
 645.0,
 200.0,
 271.0,
 1088.0,
 428.0,
 1145.0,
 496.0,
 635.0,
 921.0,
 860.0,
 1238.0,
 1117.0,
 1441.0,
 1498.0,
 1768.0,
 2308.0,
 150

In [59]:
flight_vaex.CarrierDelay.value_counts()

0.0       26679478
1.0         111661
2.0         110619
3.0         107441
6.0         104246
            ...   
2238.0           1
2109.0           1
1766.0           1
1555.0           1
2140.0           1
Length: 1884, dtype: int64

In [60]:
(26679478/30132671)*100

88.54003682580944

In [61]:
# vaex.expr.pearsonr(flight_vaex['CarrierDelay'],flight_vaex['DepDelay'])
# flight_vaex[['CarrierDelay','DepDelay']].stat.correlation()

flight_vaex.correlation('CarrierDelay','DepDelay')

array(0.69709477)

In [62]:
# Weather Delay (	Weather Delay, in Minutes)

In [63]:
flight_vaex.WeatherDelay

Expression = WeatherDelay
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  0
       1  0
       2  0
       3  0
       4  0
    ...    
30132667  0
30132668  0
30132669  0
30132670  0
30132671  0

In [64]:
len(flight_vaex.WeatherDelay.unique())

1379

In [65]:
flight_vaex.WeatherDelay.value_counts()

0.0       29746720
15.0          8181
1.0           7978
2.0           7913
6.0           7865
            ...   
1470.0           1
1301.0           1
1519.0           1
1373.0           1
1847.0           1
Length: 1379, dtype: int64

In [66]:
(29746720/30132671)*100

98.71916100633761

In [67]:
flight_vaex.correlation('WeatherDelay','DepDelay')

array(0.30413114)

WeatherDelay has 98.7 perc value 0, drop it

In [68]:
# NAS Delay (National Air System Delay, in Minutes)

In [69]:
flight_vaex.NASDelay

Expression = NASDelay
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  0
       1  0
       2  0
       3  0
       4  0
    ...    
30132667  0
30132668  0
30132669  0
30132670  0
30132671  0

In [70]:
len(flight_vaex.NASDelay.unique())

1300

In [71]:
flight_vaex.NASDelay.value_counts()

0.0       26882919
1.0         147052
15.0        132741
16.0        120433
2.0         119700
            ...   
882.0            1
1419.0           1
1145.0           1
1277.0           1
1711.0           1
Length: 1300, dtype: int64

In [72]:
(26882919/30132671)*100

89.21518772763291

In [73]:
flight_vaex.correlation('NASDelay','DepDelay')

array(0.28571692)

In [74]:
# Security Delay

In [75]:
flight_vaex.SecurityDelay

Expression = SecurityDelay
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  0
       1  0
       2  0
       3  0
       4  0
    ...    
30132667  0
30132668  0
30132669  0
30132670  0
30132671  0

In [76]:
len(flight_vaex.SecurityDelay.unique())

348

In [77]:
flight_vaex.SecurityDelay.value_counts()

0.0      30103548
15.0         1179
16.0         1047
18.0          945
17.0          936
           ...   
369.0           1
204.0           1
398.0           1
550.0           1
308.0           1
Length: 348, dtype: int64

In [78]:
(30103548/30132671)*100

99.90335075174717

In [79]:
flight_vaex.correlation('SecurityDelay','DepDelay')

array(0.03542479)

SecurityDelay Column has 99.9 perc value 0, can drop this column

In [80]:
# LateAircraftDelay

In [81]:
flight_vaex.LateAircraftDelay

Expression = LateAircraftDelay
Length: 30,132,672 dtype: float64 (column)
------------------------------------------
       0  0
       1  2
       2  0
       3  0
       4  0
    ...    
30132667  0
30132668  0
30132669  0
30132670  0
30132671  0

In [82]:
len(flight_vaex.LateAircraftDelay.unique())

1519

In [83]:
flight_vaex.LateAircraftDelay.value_counts()

0.0       26977089
15.0         75493
16.0         71538
17.0         67883
18.0         65207
            ...   
1434.0           1
1539.0           1
1432.0           1
1041.0           1
1330.0           1
Length: 1519, dtype: int64

In [84]:
(26977089/30132671)*100

89.52770565875159

In [85]:
flight_vaex.correlation('LateAircraftDelay','DepDelay')

array(0.60874058)

Droping cols Year, Month, DayofMonth,SecurityDelay,NASDelay,WeatherDelay

In [86]:
flight_vaex = flight_vaex.drop(columns=['Year','Month','DayofMonth','SecurityDelay','NASDelay','WeatherDelay'])

In [87]:
flight_vaex['FlightDate'] = flight_vaex['FlightDate'].astype('datetime64')

In [88]:
flight_vaex['Was Weekend']=  flight_vaex['FlightDate'].dt.dayofweek.isin([5,6])

In [89]:
flight_vaex.head()

#  FlightDate           Marketing_Airline_Network    OriginCityName    DestCityName      CRSDepTime    DepTime    DepDelay    DepDelayMinutes    TaxiOut    WheelsOff    WheelsOn    TaxiIn    CRSArrTime    ArrTime    ArrDelay    ArrDelayMinutes    CRSElapsedTime    ActualElapsedTime    AirTime    Distance    DistanceGroup    CarrierDelay    LateAircraftDelay  Was Weekend
  0  2018-01-15 00:00:00  UA                           Newark, NJ        Charleston, SC          1845       1928          43                 43         35         2003        2145         4          2108       2149          41                 41               143                  141        102         628                3              41                    0  False
  1  2018-01-16 00:00:00  UA                           Newark, NJ        Charleston, SC          1835       1956          81                 81         18         2014        2202         5          2058       2207          69                 69               143                  131        108         628                3              67                    2  False
  2  2018-01-17 00:00:00  UA                           Newark, NJ        Charleston, SC          1835       1836           1                  1         19         1855        2048         6          2058       2054          -4                  0               143                  138        113         628                3               0                    0  False
  3  2018-01-18 00:00:00  UA                           Newark, NJ        Charleston, SC          1845       1844          -1                  0         36         1920        2052         6          2108       2058         -10                  0               143                  134         92         628                3               0                    0  False
  4  2018-01-20 00:00:00  UA                           Newark, NJ        Charleston, SC          1835       1829          -6                  0         22         1851        2020         6          2058       2026         -32                  0               143                  117         89         628                3               0                    0  True
  5  2018-01-21 00:00:00  UA                           Newark, NJ        Charleston, SC          1835       1825         -10                  0         44         1909        2039         5          2058       2044         -14                  0               143                  139         90         628                3               0                    0  True
  6  2018-01-22 00:00:00  UA                           Newark, NJ        Charleston, SC          1845       1847           2                  2         41         1928        2118        10          2108       2128          20                 20               143                  161        110         628                3               2                    0  False
  7  2018-01-23 00:00:00  UA                           Newark, NJ        Charleston, SC          1835       1823         -12                  0         34         1857        2103        11          2058       2114          16                 16               143                  171        126         628                3               0                    0  False
  8  2018-01-25 00:00:00  UA                           Newark, NJ        Charleston, SC          1845       1841          -4                  0         38         1919        2056         7          2108       2103          -5                  0               143                  142         97         628                3               0                    0  False
  9  2018-01-26 00:00:00  UA                           Newark, NJ        Charleston, SC          1845       1838          -7                  0         39         1917        2047         6          2108       2053         -15                  0               143                  135         90         628 

Note CRS ELapsed time = CRSDep Time - CRSArr Time
,same in case for actual

In [90]:
def airline_abbreviation_to_name(x): 
    airline_abb = {
        "UA" : "United Airlines",
        "DL" : "Delta Airlines",
        "F9" : "Frontier Airlines",
        "NK" : "Spirit Airlines",
        "AA" : "American Airlines",
        "WN" : "Southwest Airlines",
        "AS" : "Alaska Airlines",
        "HA" : "Hawaiian Airlines",
        "VX" : "Virgin America",
        "B6" : "JetBlue Airways",
        "G4" : "Allegiant Air",    
    }
    for key, value in airline_abb.items():
        if x == key:
            x = value
    return x

flight_vaex["Airline"] = flight_vaex["Marketing_Airline_Network"].apply(airline_abbreviation_to_name)

In [91]:
flight_vaex = flight_vaex.drop(columns=['Marketing_Airline_Network'])

In [92]:
flight_vaex.head()

#  FlightDate           OriginCityName    DestCityName      CRSDepTime    DepTime    DepDelay    DepDelayMinutes    TaxiOut    WheelsOff    WheelsOn    TaxiIn    CRSArrTime    ArrTime    ArrDelay    ArrDelayMinutes    CRSElapsedTime    ActualElapsedTime    AirTime    Distance    DistanceGroup    CarrierDelay    LateAircraftDelay  Was Weekend    Airline
  0  2018-01-15 00:00:00  Newark, NJ        Charleston, SC          1845       1928          43                 43         35         2003        2145         4          2108       2149          41                 41               143                  141        102         628                3              41                    0  False          United Airlines
  1  2018-01-16 00:00:00  Newark, NJ        Charleston, SC          1835       1956          81                 81         18         2014        2202         5          2058       2207          69                 69               143                  131        108         628                3              67                    2  False          United Airlines
  2  2018-01-17 00:00:00  Newark, NJ        Charleston, SC          1835       1836           1                  1         19         1855        2048         6          2058       2054          -4                  0               143                  138        113         628                3               0                    0  False          United Airlines
  3  2018-01-18 00:00:00  Newark, NJ        Charleston, SC          1845       1844          -1                  0         36         1920        2052         6          2108       2058         -10                  0               143                  134         92         628                3               0                    0  False          United Airlines
  4  2018-01-20 00:00:00  Newark, NJ        Charleston, SC          1835       1829          -6                  0         22         1851        2020         6          2058       2026         -32                  0               143                  117         89         628                3               0                    0  True           United Airlines
  5  2018-01-21 00:00:00  Newark, NJ        Charleston, SC          1835       1825         -10                  0         44         1909        2039         5          2058       2044         -14                  0               143                  139         90         628                3               0                    0  True           United Airlines
  6  2018-01-22 00:00:00  Newark, NJ        Charleston, SC          1845       1847           2                  2         41         1928        2118        10          2108       2128          20                 20               143                  161        110         628                3               2                    0  False          United Airlines
  7  2018-01-23 00:00:00  Newark, NJ        Charleston, SC          1835       1823         -12                  0         34         1857        2103        11          2058       2114          16                 16               143                  171        126         628                3               0                    0  False          United Airlines
  8  2018-01-25 00:00:00  Newark, NJ        Charleston, SC          1845       1841          -4                  0         38         1919        2056         7          2108       2103          -5                  0               143                  142         97         628                3               0                    0  False          United Airlines
  9  2018-01-26 00:00:00  Newark, NJ        Charleston, SC          1845       1838          -7                  0         39         1917        2047         6          2108       2053         -15                  0               143                  135         90         628                3               0                    0  False          United Airl

In [93]:
# From FLightDate Column 
# new col "year", "DayOfWeek","DayOfMonth" ,"Month"

In [94]:
flight_vaex['Year']=flight_vaex['FlightDate'].dt.year
flight_vaex['DayOfWeek']=flight_vaex['FlightDate'].dt.dayofweek
flight_vaex['Month']=flight_vaex['FlightDate'].dt.month
flight_vaex['DayOfMonth']=flight_vaex['FlightDate'].dt.day

In [95]:
flight_vaex.head()

#  FlightDate           OriginCityName    DestCityName      CRSDepTime    DepTime    DepDelay    DepDelayMinutes    TaxiOut    WheelsOff    WheelsOn    TaxiIn    CRSArrTime    ArrTime    ArrDelay    ArrDelayMinutes    CRSElapsedTime    ActualElapsedTime    AirTime    Distance    DistanceGroup    CarrierDelay    LateAircraftDelay  Was Weekend    Airline            Year    DayOfWeek    Month    DayOfMonth
  0  2018-01-15 00:00:00  Newark, NJ        Charleston, SC          1845       1928          43                 43         35         2003        2145         4          2108       2149          41                 41               143                  141        102         628                3              41                    0  False          United Airlines    2018            0        1            15
  1  2018-01-16 00:00:00  Newark, NJ        Charleston, SC          1835       1956          81                 81         18         2014        2202         5          2058       2207          69                 69               143                  131        108         628                3              67                    2  False          United Airlines    2018            1        1            16
  2  2018-01-17 00:00:00  Newark, NJ        Charleston, SC          1835       1836           1                  1         19         1855        2048         6          2058       2054          -4                  0               143                  138        113         628                3               0                    0  False          United Airlines    2018            2        1            17
  3  2018-01-18 00:00:00  Newark, NJ        Charleston, SC          1845       1844          -1                  0         36         1920        2052         6          2108       2058         -10                  0               143                  134         92         628                3               0                    0  False          United Airlines    2018            3        1            18
  4  2018-01-20 00:00:00  Newark, NJ        Charleston, SC          1835       1829          -6                  0         22         1851        2020         6          2058       2026         -32                  0               143                  117         89         628                3               0                    0  True           United Airlines    2018            5        1            20
  5  2018-01-21 00:00:00  Newark, NJ        Charleston, SC          1835       1825         -10                  0         44         1909        2039         5          2058       2044         -14                  0               143                  139         90         628                3               0                    0  True           United Airlines    2018            6        1            21
  6  2018-01-22 00:00:00  Newark, NJ        Charleston, SC          1845       1847           2                  2         41         1928        2118        10          2108       2128          20                 20               143                  161        110         628                3               2                    0  False          United Airlines    2018            0        1            22
  7  2018-01-23 00:00:00  Newark, NJ        Charleston, SC          1835       1823         -12                  0         34         1857        2103        11          2058       2114          16                 16               143                  171        126         628                3               0                    0  False          United Airlines    2018            1        1            23
  8  2018-01-25 00:00:00  Newark, NJ        Charleston, SC          1845       1841          -4                  0         38         1919        2056         7          2108       2103          -5                  0               143                  142         97         628                3               0          

In [96]:
flight_vaex.correlation('TaxiOut', 'DepDelay')

array(0.07888809)

In [97]:
flight_vaex.correlation('TaxiIn','DepDelay')

array(0.03074851)

In [98]:
flight_vaex.correlation('TaxiOut', 'ArrDelay')

array(0.21716151)

In [99]:
flight_vaex.correlation('TaxiIn', 'ArrDelay')

array(0.11426912)

In [100]:
# Droping Taxiout, TaxiIn, DepDelayMinutes,ArrDelayMinutes,DistanceGroup, CarrierDelay, LateAircraftDelay,WheelsOn, WheelsOff,DepTime,ArrTime

In [101]:
flight_vaex = flight_vaex.drop(columns=['TaxiOut','TaxiIn','DepDelayMinutes','ArrDelayMinutes','DistanceGroup','CarrierDelay','LateAircraftDelay','WheelsOn','WheelsOff','DepTime','ArrTime', 'ActualElapsedTime'])



In [102]:
flight_vaex.column_names

['FlightDate',
 '__FlightDate',
 '__Marketing_Airline_Network',
 'OriginCityName',
 'DestCityName',
 'CRSDepTime',
 'DepDelay',
 'CRSArrTime',
 'ArrDelay',
 'CRSElapsedTime',
 'ActualElapsedTime',
 'AirTime',
 'Distance',
 'Was Weekend',
 'Airline',
 'Year',
 'DayOfWeek',
 'Month',
 'DayOfMonth']

In [103]:
flight_vaex.head(1)

#  FlightDate           OriginCityName    DestCityName      CRSDepTime    DepDelay    CRSArrTime    ArrDelay    CRSElapsedTime    ActualElapsedTime    AirTime    Distance  Was Weekend    Airline            Year    DayOfWeek    Month    DayOfMonth
  0  2018-01-15 00:00:00  Newark, NJ        Charleston, SC          1845          43          2108          41               143                  141        102         628  False          United Airlines    2018            0        1            15

In [114]:
flight_vaex = flight_vaex.drop(columns=['AirTime'])

In [136]:
from vaex.ml import OneHotEncoder


In [151]:
le = vaex.ml.LabelEncoder()

In [154]:
flight_vaex['Weekend?'] = le.fit_transform(flight_vaex['Was Weekend'])

In [155]:
flight_vaex.head()

#  FlightDate           OriginCityName              DestCityName       CRSDepTime    DepDelay    CRSArrTime    ArrDelay    CRSElapsedTime    Distance  Was Weekend    Airline              Year    DayOfWeek    Month    DayOfMonth  Weekend?
  0  2021-11-01 00:00:00  Chicago, IL                 Columbia, MO             1930          71          2101          62                91         315  False          American Airlines    2021            0       11             1  False
  1  2021-01-25 00:00:00  Detroit, MI                 Charlotte, NC            1714         -14          1908          -3               114         500  False          American Airlines    2021            0        1            25  False
  2  2022-01-09 00:00:00  Detroit, MI                 Atlanta, GA              2105           2          2318          -9               133         594  True           Delta Airlines       2022            6        1             9  True
  3  2020-03-22 00:00:00  Atlanta, GA                 Baton Rouge, LA          1643          -8          1724         -19               101         448  True           Delta Airlines       2020            6        3            22  True
  4  2019-08-05 00:00:00  Seattle, WA                 Detroit, MI              1153          11          1914          18               261        1927  False          Delta Airlines       2019            0        8             5  False
  5  2019-02-09 00:00:00  Cedar Rapids/Iowa City, IA  Denver, CO                600           0           717          28               137         692  True           United Airlines      2019            5        2             9  True
  6  2023-03-08 00:00:00  Atlanta, GA                 Wilmington, NC            902          -3          1028         -19                86         377  False          Delta Airlines       2023            2        3             8  False
  7  2018-12-01 00:00:00  Albany, NY                  Newark, NJ                820         170           934         175                74         143  True           United Airlines      2018            5       12             1  True
  8  2022-10-07 00:00:00  Houston, TX                 New York, NY             1430          -2          1913         -28               223        1416  False          United Airlines      2022            4       10             7  False
  9  2019-03-23 00:00:00  Fort Lauderdale, FL         Los Angeles, CA          2055          -5          2355          -7               360        2343  True           Spirit Airlines      2019            5        3            23  True

In [137]:
flight_vaex = flight_vaex.shuffle()
df_train, df_test = flight_vaex.ml.train_test_split(test_size=0.2)

C:\Users\HP\Desktop\Programming\Projects\Project 4- Flight\flight\lib\site-packages\vaex\ml\__init__.py:28: UserWarning: Make sure the DataFrame is shuffled
  warnings.warn('Make sure the DataFrame is shuffled')


In [138]:
encoder=vaex.ml.OneHotEncoder()

In [149]:
df_train['Airline'] = encoder.fit_transform(df_train['Airline'])


In [150]:
df_train.head()

#  FlightDate           OriginCityName      DestCityName        CRSDepTime    DepDelay    CRSArrTime    ArrDelay    CRSElapsedTime    Distance  Was Weekend    Airline               Year    DayOfWeek    Month    DayOfMonth
  0  2019-03-26 00:00:00  Washington, DC      Atlanta, GA                600          -7           754         -25               114         547  False          Delta Airlines        2019            1        3            26
  1  2018-02-16 00:00:00  Indianapolis, IN    New York, NY              1147          35          1348          36               121         660  False          Delta Airlines        2018            4        2            16
  2  2022-03-20 00:00:00  Orlando, FL         Austin, TX                 710          -4           910         -15               180         994  True           Southwest Airlines    2022            6        3            20
  3  2021-07-04 00:00:00  San Jose, CA        Phoenix, AZ               1930           2          2120          -8               110         621  True           Southwest Airlines    2021            6        7             4
  4  2021-12-24 00:00:00  Chicago, IL         Charlotte, NC             1425          -2          1715         -23               110         584  False          Southwest Airlines    2021            4       12            24
  5  2018-07-14 00:00:00  Chicago, IL         Los Angeles, CA           1944          66          2211          29               267        1744  True           United Airlines       2018            5        7            14
  6  2022-05-23 00:00:00  Phoenix, AZ         Austin, TX                1000           2          1415          -5               135         872  False          Southwest Airlines    2022            0        5            23
  7  2023-03-02 00:00:00  Atlanta, GA         Baltimore, MD             2230          -1            10          -9               100         577  False          Southwest Airlines    2023            3        3             2
  8  2018-12-16 00:00:00  Atlanta, GA         Myrtle Beach, SC           941          -3          1048          -5                67         317  True           Delta Airlines        2018            6       12            16
  9  2018-02-07 00:00:00  Raleigh/Durham, NC  Newark, NJ                1730          79          1921          54               111         416  False          United Airlines       2018            2        2             7

In [141]:
df_test = encoder.transform(df_test)

In [144]:
df_train.head()

#  FlightDate           OriginCityName      DestCityName        CRSDepTime    DepDelay    CRSArrTime    ArrDelay    CRSElapsedTime    Distance  Was Weekend    Airline               Year    DayOfWeek    Month    DayOfMonth
  0  2019-03-26 00:00:00  Washington, DC      Atlanta, GA                600          -7           754         -25               114         547  False          Delta Airlines        2019            1        3            26
  1  2018-02-16 00:00:00  Indianapolis, IN    New York, NY              1147          35          1348          36               121         660  False          Delta Airlines        2018            4        2            16
  2  2022-03-20 00:00:00  Orlando, FL         Austin, TX                 710          -4           910         -15               180         994  True           Southwest Airlines    2022            6        3            20
  3  2021-07-04 00:00:00  San Jose, CA        Phoenix, AZ               1930           2          2120          -8               110         621  True           Southwest Airlines    2021            6        7             4
  4  2021-12-24 00:00:00  Chicago, IL         Charlotte, NC             1425          -2          1715         -23               110         584  False          Southwest Airlines    2021            4       12            24
  5  2018-07-14 00:00:00  Chicago, IL         Los Angeles, CA           1944          66          2211          29               267        1744  True           United Airlines       2018            5        7            14
  6  2022-05-23 00:00:00  Phoenix, AZ         Austin, TX                1000           2          1415          -5               135         872  False          Southwest Airlines    2022            0        5            23
  7  2023-03-02 00:00:00  Atlanta, GA         Baltimore, MD             2230          -1            10          -9               100         577  False          Southwest Airlines    2023            3        3             2
  8  2018-12-16 00:00:00  Atlanta, GA         Myrtle Beach, SC           941          -3          1048          -5                67         317  True           Delta Airlines        2018            6       12            16
  9  2018-02-07 00:00:00  Raleigh/Durham, NC  Newark, NJ                1730          79          1921          54               111         416  False          United Airlines       2018            2        2             7

In [116]:
df_train.shape

(24106138, 15)

In [117]:
df_test.shape

(6026534, 15)

In [131]:
from vaex.ml.sklearn import IncrementalPredictor
# from sklearn.multioutput import MultiOutputRegressor
from sklearn.linear_model import SGDRegressor

In [130]:
features = ['OriginCityName', 'DestCityName', 'CRSDepTime', 'CRSArrTime', 'CRSElapsedTime','Distance', 'Was Weekend', 'Airline', 'Year', 'DayOfWeek', 'Month', 'DayOfMonth']
target = 'DepDelay'

In [132]:
model = SGDRegressor()

vaex_model = IncrementalPredictor(features=features, target=target,model=model, batch_size=500000)

In [ ]:
vaex_model.fit(df=df_train,progress='widget')